src/wp-includes/post.php

```diff
@@ -5690,45 +5690,63 @@ function get_page( $page, $output = OBJECT, $filter = 'raw' ) {
-	$cache_key = "get_page_by_path:$hash:$last_changed";
-	$cached    = wp_cache_get( $cache_key, 'posts' );
-	if ( false !== $cached ) {
-		// Special case: '0' is a bad `$page_path`.
-		if ( '0' === $cached || 0 === $cached ) {
-			return;
-		} else {
-			return get_post( $cached, $output );
+	$use_wp_query = ! has_filter( 'posts_pre_query' ) && ! has_action( 'pre_get_posts' );
+
+	if ( ! $use_wp_query ) {
+		$last_changed = wp_cache_get_last_changed( 'posts' );
+		$hash         = md5( $page_path . serialize( $post_type ) );
+		$cache_key    = "get_page_by_path:$hash:$last_changed";
+		$cached       = wp_cache_get( $cache_key, 'posts' );
+		if ( false !== $cached ) {
+			// Special case: '0' is a bad `$page_path`.
+			if ( '0' === $cached || 0 === $cached ) {
+				return;
+			} else {
+				return get_post( $cached, $output );
+			}
 		}
 	}
 
-	$page_path     = rawurlencode( urldecode( $page_path ) );
-	$page_path     = str_replace( '%2F', '/', $page_path );
-	$page_path     = str_replace( '%20', ' ', $page_path );
-	$parts         = explode( '/', trim( $page_path, '/' ) );
-	$parts         = array_map( 'sanitize_title_for_query', $parts );
-	$escaped_parts = esc_sql( $parts );
-
-	$in_string = "'" . implode( "','", $escaped_parts ) . "'";
+	$page_path = rawurlencode( urldecode( $page_path ) );
+	$page_path = str_replace( '%2F', '/', $page_path );
+	$page_path = str_replace( '%20', ' ', $page_path );
+	$parts     = explode( '/', trim( $page_path, '/' ) );
+	$parts     = array_map( 'sanitize_title_for_query', $parts );
 
 	if ( is_array( $post_type ) ) {
 		$post_types = $post_type;
 	} else {
 		$post_types = array( $post_type, 'attachment' );
 	}
 
-	$post_types          = esc_sql( $post_types );
-	$post_type_in_string = "'" . implode( "','", $post_types ) . "'";
-	$sql                 = "
-		SELECT ID, post_name, post_parent, post_type
-		FROM $wpdb->posts
-		WHERE post_name IN ($in_string)
-		AND post_type IN ($post_type_in_string)
-	";
+	if ( $use_wp_query ) {
+		$args = array(
+			'post_name__in'          => $parts,
+			'post_type'              => $post_types,
+			'post_status'            => 'all',
+			'posts_per_page'         => -1,
+			'update_post_term_cache' => false,
+			'update_post_meta_cache' => false,
+			'no_found_rows'          => true,
+			'suppress_filters'       => true,
+			'orderby'                => 'none',
+		);
+
+		$query = new WP_Query( $args );
+		$posts = $query->get_posts();
+		$pages = array();
 
-	$pages = $wpdb->get_results( $sql, OBJECT_K );
+		foreach ( $posts as $post ) {
+			$pages[ $post->ID ] = $post;
+		}
+	} else {
+		$escaped_parts       = esc_sql( $parts );
+		$in_string           = "'" . implode( "','", $escaped_parts ) . "'";
+		$post_types          = esc_sql( $post_types );
+		$post_type_in_string = "'" . implode( "','", $post_types ) . "'";
+		$sql                 = "SELECT ID, post_name, post_parent, post_type FROM $wpdb->posts WHERE post_name IN ($in_string) AND post_type IN ($post_type_in_string)";
```

Comment:
```suggestion
		$sql                 = "
			SELECT ID, post_name, post_parent, post_type
			FROM $wpdb->posts
			WHERE post_name IN ($in_string)
			AND post_type IN ($post_type_in_string)
		";
```
Minor nitpick, I find the existing wrapping into multiple lines more readable than one long line.